### ISSUE TITLE
### Summary of Changes

### ISSUE THREAD

**Issue by chaichanachai54-bit** [apple/coreai-torch#90]
### Summary of Changes

Closes #94.

This PR resolves two production issues in `KMeansPalettizer`:

1. **Deterministic Checkpoint Cleanup & State Rollback in `calibration_mode`**:
   - Refactored `calibration_mode` context manager to cleanly separate the normal commit path from the error rollback path (`finally:`).
   - Ensured that if an exception is raised inside `with palettizer.calibration_mode(...)`:
     - The caller's original exception propagates with its full traceback rather than being masked by a secondary `RuntimeError`.
     - The temporary `.pt` model checkpoint file is guaranteed to be unlinked, preventing disk leaks in `/tmp`.
     - Model weights are rolled back from the checkpoint to their pre-calibration state.
     - Fake palettization is re-enabled (`self._model.apply(_enable_fake_palett)`), ensuring the model is not left in a disabled state.
     - Compressor lifecycle resets to `IDLE`.
   - In `_load_model_checkpoint()`, wrapped `os.unlink` in `try ... finally` to guarantee cleanup even if `torch.load` / `load_state_dict` encounters deserialization errors.

2. **Root Module Parameter Name Normalization**:
   - In `_set_sensitivities_in_fake_palettize_modules` and `_get_sensitivities_from_fake_palettize_modules`, updated parameter name construction to:
     ```python
     param_name = ".".join(filter(None, [module_name, "parametrizations", attr_name, "original"]))
     ```
     This eliminates the invalid leading dot when `module_name == ""` (root module), matching PyTorch's `named_parameters()` key hierarchy (`"parametrizations.weight.original"`).

---

### Motivation & Impact

In large-scale or multi-job model optimization pipelines:
* Model checkpoints saved during calibration can range from hundreds of megabytes to tens of gigabytes. Any aborted run on `main` orphaned these files in `/tmp`, leading to disk space exhaustion.
* Masking caller exceptions hindered debugging by suppressing the root cause traceback.
* Standalone root modules (e.g. single linear/embedding blocks being optimized) failed to receive sensitivities due to leading-dot key mismatch.

---

### Automated Tests Added

Added to `tests/palettization/test_kmeans_palettizer.py`:
* `test_calibration_mode_exception_cleans_up_checkpoint_and_restores_state`:
  - Validates pre-step abort: caller exception is unmasked, temp checkpoint is unlinked, fake palettization is re-enabled, and lifecycle returns to `IDLE`.
  - Validates post-step abort: mutated weights are rolled back to pre-calibration state, temp checkpoint is unlinked, fake palettization is re-enabled, and lifecycle returns to `IDLE`.
  - Uses a process-safe spy pattern on `_save_model_checkpoint` to prevent race conditions during parallel test execution (`pytest -n auto`).
* `test_calibration_mode_root_module`:
  - Validates that a standalone root `nn.Linear` module completes calibration and successfully assigns normalized sensitivity buffers to `_KMeansFakePalettize`.

---

### Verification

* `pytest tests/palettization/test_kmeans_palettizer.py`: **83 passed in 7.59s**
* `pytest tests/palettization/ -m "not slow" -n auto`: **420 passed in 14.30s**
* `make check`: **All checks passed** (`ruff`, `darker`, license header, towncrier check, etc.)

---

### Checklist
- [x] Follows Conventional Commits formatting
- [x] Commit is cryptographically signed with SSH key
- [x] Includes Towncrier fragment (`changelog.d/94.fixed`)
- [x] Passes `make check` without warnings or diff pollution
- [x] Includes automated tests reproducing the bug and proving the fix

_Originally posted by @rohith500 in https://github.com/apple/coreai-optimization/pull/95_
